### PR TITLE
Fix the order of workflow context update

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Issue #436 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

The order that workflow dispatcher updates the workflow status has been changed.
If execution delay is set and job is Inflight, the context will get updated.
An integration test has been added.

### Tests

- [x] The following tests are written for this issue:
TestStopWorkflowWithExecutionDelay

- [x] The following is the result of the "mvn test" command on the appropriate module:

Test Result 1: mvn test
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 846, Failures: 2, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  56:03 min
[INFO] Finished at: 2019-09-03T13:26:43-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/anajari/my_repos/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

Test Result 2: mvn test -Dtest="TestAlertingRebalancerFailure,TestTaskPerformanceMetrics"
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 19.496 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.721 s
[INFO] Finished at: 2019-09-03T13:49:34-07:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

**Comment by narendly** [comment]
@alirezazamani Could you please paste the logs of failed test from now on so we have some logs to go off of when we fix tests?

**Comment by alirezazamani** [comment]
@narendly Sure.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.helix.task.Workflow;
+import org.apache.helix.task.WorkflowContext;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Test To check if the Execution Delay is respected.
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java]: Let's add more description here. There is more you're testing - not only are you testing whether execution delay is respected, you're also testing whether stop works properly with execution delay set. Please help the reader understand why you wrote this test.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+  }
+
+  @Test
+  public void testStopWorkflowWithExecutionDelay() throws Exception {
+    // Execution Delay is set to be 20 milliseconds.
+    final long executionDelay = 20L;
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java]: Is there a reason why this is set at 20L? Would it make a difference if it was 200, 2000? Or does any delay do it? Just curious. Might be good to clarify here.

Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java]: No there is no reason to use 20L. As long as we use any execution delay and the job goes to Inflight job queue, we need this fix. I will clarify on the code.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestStopWorkflowWithExecutionDelay.java]: Added. Thanks.

**Comment by alirezazamani** [comment]
This PR is ready to be merged. Approved by @narendly 

Final Commit message:

Title: 
Fix the order of workflow context update
    
Description:
In this commit:
The order that workflow dispatcher updates the workflow status has been changed.
If execution delay is set and job is inflight, the context will get updated.
An integration test has been added.

**Pull request merged**